### ISSUE TITLE
Unable to pull screenshot baseline

### ISSUE THREAD

**Issue by isuPatches** [ndtp/android-testify#126]
**Describe the bug**

Unable to pull screenshot baseline and witnessing errors in the logs.  I may be doing something incorrectly though 😅 .

This issue relates to:
- [ ] The Kotlin library
- [x] The Gradle plugin. (at least I think...)
- [ ] The IntelliJ Platform plugin
- [ ] The sample code
- [ ] The documentation

**To Reproduce**
1. I am running `./gradlew module:screenshotRecord`
2. I am then running `./gradlew module:screenshotPull`

**Expected behavior**
- No errors in the logs
- Screenshots copied from device to `androidTest/assets/screenshots`

**Acutal behavior**
- Errors in the logs about not being able to locate a *.jar file
- Errors in the logs about files not being found
- No screenshots in `androidTest/assets/screenshots`

What I have observed in the logs:

```kotlin
2023-05-01 10:19:22.582 31079-31118 System                  com.mypackage.test  E  Unable to open zip file: /data/user/0/com.mypackage.test/cache/vtt7tgqF.jar
2023-05-01 10:19:22.584 31079-31118 System                  com.mypackage.test  E  java.io.FileNotFoundException: /data/user/0/com.mypackage.test/cache/vtt7tgqF.jar (No such file or directory)                                                                                               	
```

```kotlin
2023-05-01 10:19:23.358 31079-31118 AssetLoader              com.mypackage.test  E  Unable to decode file screenshots/33-1080x2400@450dp-en_US/MyTestClass_test.png.
                                                                                                    java.io.FileNotFoundException: screenshots/33-1080x2400@450dp-en_US/MyTestClass_test.png
```

**Screenshots**

N/A

**Desktop (please complete the following information):**
 - OS: macOS
 - Version: 13.3.1

**Target Android Device (please complete the following information):**
 - Device: Pixel 6a, Samsung S21+
-  Physical or Virtual: Physical
 - API Level: 33
- Testify Key: 

```kotlin
  Format: {api_version}-{width_in_pixels}x{height_in_pixels}@{dpi}_{locale}
  key                  = 33-1080x2400@450dp-en_US
```

**Additional context**

I have also tried on various emulators and with various permissions settings. 

I am on:
- AGP 8
- Kotlin 1.8.20
- Android Studio Flamingo
- Java 11
- I am using compose and `androidTestImplementation("dev.testify:testify-compose:2.0.0-beta01")`

**Comment by DanielJette** [comment]
@isuPatches 

🤔 I've not encountered these errors before. This is perplexing 

The `FileNotFoundException` on the `jar` file reads like it could be a permission issue reading the assets file (the baseline images).

If you're able, there are a couple of things you could try to help narrow down the cause of the issues.

- You could try to configure an emulator according to these instructions: https://ndtp.github.io/android-testify/docs/get-started/configuring-an-emulator
   It's possible there are some permission changes in the Android SDK that Testify is not handling correctly. Using an older API level like the one in the documentation could narrow down the cause of the errors
- You can try a few running a few of the other gradle tasks to see if the Gradle Plugin is causing issues
    `.gradlew module:testifySettings` would be a good one to start with
- Can you try running the tests directly from Android Studio? Use the ▶️ Run action next to the test method to run a screenshot test as a normal androidTest Instrumentation test. This will skip the Testify Gradle Plugin and eliminate one of the variables

**Comment by isuPatches** [comment]
👋🏽  I followed what you mentioned above.

- With an emulator setup per the docs (SDK 29, Pixel 3a) I saw the same errors in the logs
- With the other testify commands:

```
> Task :module:testifySettings

------------------------------------------------------------
Displays the Testify gradle extension settings
------------------------------------------------------------

  baselineSourceDir      = /Users/patches/Documents/GitHub/android-mypackage/module/src/androidTest/assets
  installAndroidTestTask = installDebugAndroidTest
  installTask            = null
  moduleName             = module
  outputFileNameFormat   = null
  pullWaitTime           = 0
  targetPackageId        = com.mypackage
  testPackageId          = com.mypackage.test
  testRunner             = com.mypackage.HiltTestRunner
  useSdCard              = false
```

```
> Task :module:testifyVersion

------------------------------------------------------------
Displays the Testify plugin version
------------------------------------------------------------

  Vendor               = ndtp
  Title                = Testify
  Version              = 2.0.0-beta01
```

```
> Task :module:deviceLocale

------------------------------------------------------------
Displays the device locale.
------------------------------------------------------------

  Current Locale       = en_US

```

and other reasonable output from the command line.


- I was unable to run the plugin from Android Studio and received this error with Java 11:

```
Build file '/Users/patches/Documents/GitHub/android-mypackage/app/build.gradle.kts' line: 10

An exception occurred applying plugin request [id: 'com.android.application']
> Failed to apply plugin 'com.android.internal.application'.
   > Android Gradle plugin requires Java 17 to run. You are currently using Java 11.
      Your current JDK is located in /Library/Java/JavaVirtualMachines/jdk-11.0.12.jdk/Contents/Home
      You can try some of the following options:
       - changing the IDE settings.
       - changing the JAVA_HOME environment variable.
       - changing `org.gradle.java.home` in `gradle.properties`.
```

With Java 17:

```
Cannot locate tasks that match ':appfeatures.module.android-mypackage.appfeatures.module:screenshotRecord' as project 'appfeatures.module.android-mypackage.appfeatures.module' not found in root project 'android-mypackage'.

* Try:
> Run gradle projects to get a list of available projects.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
```

I believe this is an unrelated issue though as I have seen this when a project is setup as:

app
appfeatures
 |---module1
 |---module2

I believe the inability for the screenshots to be pulled could be a permissions issue as you mentioned but that the Intellij plugin part is hiccuping on nested modules as well.

Do you think it would be worth me trying to downgrade Android Studio tomorrow after work?  Curious if Flamingo shipping with Java 17 has caused this hiccup.

**Comment by DanielJette** [comment]
Thanks for providing all of this information @isuPatches 

From your information, I think you're dealing with two issues: the JDK and multi-module support.

**JDK**

I've been able to get Testify working correct with Flamingo and Java 17. I think you need to update your Android Studio settings for this.

I believe you want to [set the JDK version](https://developer.android.com/studio/intro/studio-config#jdk) to `Embedded JDK`:
<img width="981" alt="image" src="https://user-images.githubusercontent.com/5921367/235680142-c17071f1-454a-4e57-8b37-66005a6b125a.png">

**Multi-Module Support**

Thank you for providing the output from `:module:testifySettings`.

These two values are of interest to me:

```
  targetPackageId        = com.mypackage
  testPackageId          = com.mypackage.test
```

I suspect that these are incorrect. Is it possible that those are the values for your app, not the module you're testing?
If you have added Testify to the app module as well, then this would mean that the commands would "work", but the paths for pulling and permission for accessing the data files would be wrong. It could explain the errors you're seeing.

The configuration for module projects is currently less than ideal. There are plans to improve the documentation and auto-detection of settings, but for now, may I suggest you try to change the `testify` settings in your `build.gradle` file?

For module projects, you want the `applicationPackageId` and the `testPackageId` to be the same. An Android library project will generate a single apk (usually named `module-debug-androidTest.apk`) which contains both your module code and the test harness. So, you need to tell Testify that the "application" under test is actually the generated androidTest apk.

For example, if your module's package is `com.example.module`, then you'd want to set the `testify` settings to this:

```
testify {
    applicationPackageId "com.example.module.test"
    testPackageId "com.example.module.test"
}
```

Can you try updating these settings and see if that makes any difference?

**Comment by isuPatches** [comment]
@DanielJette 

You are a rock star!  I was misconfigured within my `build.gradle.kts` of the module:

🔴 
```kotlin
    testify {
        applicationPackageId = "com.mypackage.module"
        testPackageId = "com.mypackage.module.test"
    }
```

🟢 
```kotlin
      testify {
        applicationPackageId = "com.mypackage.module.test"
        testPackageId = "com.mypackage.module.test"
    }
```

This resolved the Multi-Module Support item.

I am still unable to use the Intellij / Android Studio plugin as it is giving this error still even with the embedded JDK set:

```
Cannot locate tasks that match ':appfeatures.module.android-mypackage.appfeatures.module:screenshotRecord' as project 'appfeatures.module.android-mypackage.appfeatures.module' not found in root project 'android-mypackage'.

* Try:
> Run gradle projects to get a list of available projects.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
```

I was able to get screenshot tests running though so I am unblocked.  I really appreciate it.  Please let me know if you want me to pull out the Intellij plugin portion to a separate issue.

**Comment by DanielJette** [comment]
🎉 I'm really glad that worked out! @isuPatches 

Regarding the Intellij / Android Studio plugin, yes, please could you open a new issue?
I'd like to get more details on that problem.

**Comment by isuPatches** [comment]
Closing this as resolved.
